src/main/kotlin/org/rust/lang/core/psi/ext/RsAbstractable.kt

```diff
@@ -6,7 +6,54 @@
 package org.rust.lang.core.psi.ext
 
 import com.intellij.psi.PsiNameIdentifierOwner
+import org.rust.lang.core.psi.*
 
 interface RsAbstractable : RsNamedElement, PsiNameIdentifierOwner {
     val isAbstract: Boolean
 }
+
+sealed class RsAbstractableOwner {
+    object Free : RsAbstractableOwner()
+    object Foreign : RsAbstractableOwner()
+    class Trait(val trait: RsTraitItem) : RsAbstractableOwner()
+    class Impl(val impl: RsImplItem, val isInherent: Boolean) : RsAbstractableOwner()
+
+    val isInherentImpl: Boolean get() = this is Impl && isInherent
+    val isTraitImpl: Boolean get() = this is Impl && !isInherent
+    val isImplOrTrait: Boolean get() = this is Impl || this is Trait
+}
+
+val RsAbstractable.owner: RsAbstractableOwner
+    get() {
+        val stubOnlyParent = when (this) {
+            is RsConstant -> if (stub != null) stub.parentStub.psi else parent
+            is RsFunction -> if (stub != null) stub.parentStub.psi else parent
+            is RsTypeAlias -> if (stub != null) stub.parentStub.psi else parent
+            else -> error("unreachable")
+        }
+        return when (stubOnlyParent) {
+            is RsForeignModItem -> RsAbstractableOwner.Foreign
+            is RsMembers -> {
+                val traitOrImpl = parent.parent
+                when (traitOrImpl) {
+                    is RsImplItem -> RsAbstractableOwner.Impl(traitOrImpl, isInherent = traitOrImpl.traitRef == null)
+                    is RsTraitItem -> RsAbstractableOwner.Trait(traitOrImpl)
+                    else -> error("unreachable")
+                }
+            }
+            else -> RsAbstractableOwner.Free
```

Comment:
`is RsItemsOwner -> RsConstantOwner.Free` from `RsConstant` is now handled by the `else` just like already done in `RsFunction`